YNU-354: Protocol Docs (#69)
layer-3/docs: docs/protocol/off-chain/channel-methods.mdx

---
sidebar_position: 4
title: Channel Management Methods
---

import Tooltip from '@site/src/components/Tooltip';

# Channel Management Methods

Channel management methods enable clients to create, modify, and close payment channels with <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">a clearnode</Tooltip> on various blockchain networks.

---

## Overview

Payment channels are the foundation of the Nitrolite protocol. They lock funds on-chain while enabling instant off-chain operations within a <Tooltip content="An abstraction that aggregates a user's funds across multiple blockchain networks, managed by Clearnode">unified balance</Tooltip>.

### Channel Lifecycle Summary

```mermaid
stateDiagram-v2
    [*] --> Requesting: create_channel (off-chain)
    Requesting --> OnChain: User submits create() transaction
    OnChain --> INITIAL: Contract locks user funds

>     INITIAL --> ACTIVE: Clearnode joins with join()

Reviewer: Outdated, Clearnode doesn't use `join` method anymore

Reviewer: So channel becomes `ACTIVE` right after `OnChain`

Author: [8211a936738924e9e67a3e56916ec7c6fee618be](https://github.com/layer-3/docs/pull/69/commits/8211a936738924e9e67a3e56916ec7c6fee618be)

    ACTIVE --> Resizing: resize_channel (optional)

>     Resizing --> ACTIVE: Atomic reset() completes

Reviewer: Not sure what is `reset()` here

Author: [8211a936738924e9e67a3e56916ec7c6fee618be](https://github.com/layer-3/docs/pull/69/commits/8211a936738924e9e67a3e56916ec7c6fee618be)

    ACTIVE --> Closing: close_channel (cooperative)
    Closing --> [*]: Funds distributed

    ACTIVE --> Disputing: challenge() (non-cooperative)
    Disputing --> [*]: Challenge period → close()

```

---

## create_channel

### Name

`create_channel`

### Usage

> Initiates the creation of a payment <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates">channel</Tooltip> between user and <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">a clearnode</Tooltip> on a specific blockchain. <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">The clearnode</Tooltip> validates the request, generates a <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates">channel</Tooltip> configuration with a unique <Tooltip content="Unique number ensuring channel identifier uniqueness">nonce</Tooltip>, prepares the initial funding state with the CHANOPEN magic number, and signs it. The user receives the complete <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates">channel</Tooltip> data and <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">the clearnode's</Tooltip> signature, which they must then submit to the blockchain's <Tooltip content="The main on-chain contract implementing channel creation, joining, closure, and resizing. It is an implementation of the Nitrolite protocol">Custody contract</Tooltip> via the `create()` function to finalize <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates">channel</Tooltip> creation and lock funds on-chain. This two-step process (off-chain preparation, on-chain execution) ensures <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">the clearnode</Tooltip> is ready to join before the user commits funds.

Reviewer: ```suggestion
Initiates the creation of a payment <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates">channel</Tooltip> between user and <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">a clearnode</Tooltip> on a specific blockchain. <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">The clearnode</Tooltip> validates the request, generates a <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates">channel</Tooltip> configuration with a unique <Tooltip content="Unique number ensuring channel identifier uniqueness">nonce</Tooltip>, prepares the initial funding state with the CHANOPEN magic number, and signs it. The user receives the complete <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates">channel</Tooltip> data and <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">the clearnode's</Tooltip> signature, which they must then submit to the blockchain's <Tooltip content="The main on-chain contract implementing channel creation, joining, closure, and resizing. It is an implementation of the Nitrolite protocol">Custody contract</Tooltip> via the `create()` function to finalize <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates">channel</Tooltip> creation and lock funds on-chain. This two-step process (off-chain preparation, on-chain execution) ensures <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">the clearnode</Tooltip> has agreed on channel creation and received an on-chain confirmation that it was created.
```

Reviewer: Changed only the last sentence

Author: [8211a936738924e9e67a3e56916ec7c6fee618be](https://github.com/layer-3/docs/pull/69/commits/8211a936738924e9e67a3e56916ec7c6fee618be)

### When to Use

> When a user wants to deposit funds and establish a payment <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates">channel</Tooltip> on a specific blockchain network. This is typically the first operation after authentication.

Reviewer: It will be potentially the first operation if the user doesn't have an open channel yet, **BUT** on subsequent connections he won't need to open the channel again unless he decided to close it.

Author: [8211a936738924e9e67a3e56916ec7c6fee618be](https://github.com/layer-3/docs/pull/69/commits/8211a936738924e9e67a3e56916ec7c6fee618be)

:::info Two-Step Process
Channel creation is intentionally split into two steps:

> 1. **Off-chain preparation**: <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">The clearnode</Tooltip> prepares and signs channel configuration

Reviewer: ```suggestion
1. **Off-chain preparation**: <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">The clearnode</Tooltip> prepares and signs the initial state
```

Author: [8211a936738924e9e67a3e56916ec7c6fee618be](https://github.com/layer-3/docs/pull/69/commits/8211a936738924e9e67a3e56916ec7c6fee618be)

2. **On-chain execution**: User submits transaction to lock funds

This ensures <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">the clearnode</Tooltip> is ready to join before the user risks their funds.
:::

### Prerequisites

- User must be [authenticated](./authentication) (<Tooltip content="A temporary cryptographic key delegated by a user's main wallet that provides a flexible way for the user to manage security of their funds by giving specific permissions and allowances for specific apps">session key</Tooltip> registered)
- User must have sufficient token balance in their wallet
- Target blockchain and token must be supported by <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">the clearnode</Tooltip>
- User must have native currency for gas fees

### Request

> | Parameter | Type | Required | Description | Default | Example | Notes |

Reviewer: I don't know whether you can do something with it, but table is not rendered in a pretty way. `description` is the longest value in each record, but more width is booked by `example`.

Author: [8211a936738924e9e67a3e56916ec7c6fee618be](https://github.com/layer-3/docs/pull/69/commits/8211a936738924e9e67a3e56916ec7c6fee618be)

|-----------|------|----------|-------------|---------|---------|-------|

> | `chain_id` | uint64 | Yes | Blockchain network identifier<br/><br/>**Examples**:<br/>• 1: Ethereum Mainnet<br/>• 137: Polygon<br/>• 8453: Base<br/>• 42161: Arbitrum One<br/>• 10: Optimism | - | `137` | Use get_config to see supported chains |

Reviewer: ```suggestion
| `chain_id` | uint32 | Yes | Blockchain network identifier<br/><br/>**Examples**:<br/>• 1: Ethereum Mainnet<br/>• 137: Polygon<br/>• 8453: Base<br/>• 42161: Arbitrum One<br/>• 10: Optimism | - | `137` | Use get_config to see supported chains |
```

Author: [8211a936738924e9e67a3e56916ec7c6fee618be](https://github.com/layer-3/docs/pull/69/commits/8211a936738924e9e67a3e56916ec7c6fee618be)

| `token` | string (wallet address) | Yes | ERC-20 token contract address on the specified chain<br/><br/>**Format**: 0x-prefixed hex string (20 bytes) | - | `"0x2791Bca1f2de4661ED88A30C99A7a9449Aa84174"` (USDC on Polygon) | Must be a supported token; see get_assets for available tokens |

> | `amount` | string | Yes | Initial deposit amount in token's smallest unit<br/><br/>**Format**: Decimal string (not hex) | - | `"100000000"` represents 100 USDC (100 * 10^6) | Must account for token decimals (e.g., USDC has 6 decimals) |

Reviewer: In the upcoming protocol version there is no such param anymore

Author: [8211a936738924e9e67a3e56916ec7c6fee618be](https://github.com/layer-3/docs/pull/69/commits/8211a936738924e9e67a3e56916ec7c6fee618be)

> | `challenge` | uint64 | No | Challenge period duration in seconds | 86400 (24 hours) | `3600` (1 hour) | Longer periods are safer but slower to close in disputes<br/>Recommended: Use default unless specific requirements |

Reviewer: There is no such param

Author: [8211a936738924e9e67a3e56916ec7c6fee618be](https://github.com/layer-3/docs/pull/69/commits/8211a936738924e9e67a3e56916ec7c6fee618be)

### Response

:::tip Quick Reference
Structures: [Channel](#channel-structure) • [State](#state-structure) • [StateAllocation](#stateallocation)
:::

| Parameter | Type | Description | See Also |
|-----------|------|-------------|----------|
| `channel` | Channel | Complete channel configuration to be submitted on-chain | [↓ Structure](#channel-structure) |
| `state` | State | Initial funding state with the clearnode's signature | [↓ Structure](#state-structure) |

> | `broker_signature` | string | The clearnode's signature over the channel and state (0x-prefixed hex, 65 bytes) | — |

Reviewer: ```suggestion
| `server_signature` | string | The clearnode's signature over the channel and state (0x-prefixed hex, 65 bytes) | — |
```

Author: [8211a936738924e9e67a3e56916ec7c6fee618be](https://github.com/layer-3/docs/pull/69/commits/8211a936738924e9e67a3e56916ec7c6fee618be)

| `channel_id` | string | Computed channel identifier: `keccak256(abi.encode(channel))` (0x-prefixed hex, 32 bytes)<br/>Client should compute this independently to verify | — |

#### Channel {#channel-structure}

| Field | Type | Description | Notes |
|-------|------|-------------|-------|
| `participants` | wallet address[] | `[User wallet address, Clearnode wallet address]` | Order: Index 0 = User, Index 1 = Clearnode<br/>Order is critical for signature verification |
| `adjudicator` | wallet address | Adjudicator contract address for this channel | Typically SimpleConsensus for payment channels<br/>Validates state transitions during disputes |
| `challenge` | uint64 | Challenge period in seconds | Matches request or default |
| `nonce` | uint64 | Unique identifier for this channel | Ensures channelId uniqueness even with same participants<br/>Server-generated timestamp or counter |

**Example**:
```json
{
  "participants": ["0x742d35Cc...", "0x123456Cc..."],
  "adjudicator": "0xAdjudicator123...",
  "challenge": 86400,
  "nonce": 1699123456
}
```

> #### State {#state-structure}

Reviewer: add `intent` and `version`

Author: [8211a936738924e9e67a3e56916ec7c6fee618be](https://github.com/layer-3/docs/pull/69/commits/8211a936738924e9e67a3e56916ec7c6fee618be)

}
```

#### StateAllocation {#stateallocation}

| Field | Type | Description |
|-------|------|-------------|
| `destination` | wallet address | Participant's wallet address |
| `token` | wallet address | Token contract address |
| `amount` | string | Amount in smallest unit (e.g., `"100000000"` for 100 USDC with 6 decimals) |

:::tip Clearnode Signature First
<Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">The clearnode</Tooltip> provides its signature BEFORE the user commits funds on-chain. This ensures both parties have committed to the channel before any on-chain transaction occurs.
:::

### Next Steps After Receiving Response

1. **Verify Channel Data**
   - Recompute `channelId` = `keccak256(abi.encode(channel))`
   - Verify computed ID matches response `channel_id`
   - Check participants[0] is your wallet address
   - Verify token and amount match your request

2. **Verify the Clearnode's Signature**
   - Compute `stateHash` = `keccak256(abi.encode(channelId, state.intent, state.version, state.data, state.allocations))`

>    - Recover signer from `broker_signature`

Reviewer: ```suggestion
   - Recover signer from `server_signature`
```

Author: [8211a936738924e9e67a3e56916ec7c6fee618be](https://github.com/layer-3/docs/pull/69/commits/8211a936738924e9e67a3e56916ec7c6fee618be)

   - Verify signer is <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">the clearnode's</Tooltip> known wallet address

3. **Sign State with Your Key**
   - Sign `stateHash` with your <Tooltip content="An entity (identified by a wallet address) that is part of a channel">participant</Tooltip> key
   - Add your signature at `state.sigs[0]`

> 4. **Approve Token Spending** (if needed)

Reviewer: It cannot be needed on this stage in the current version of the protocol as this action won't do any funds movement

Author: [8211a936738924e9e67a3e56916ec7c6fee618be](https://github.com/layer-3/docs/pull/69/commits/8211a936738924e9e67a3e56916ec7c6fee618be)

   - Check token allowance: `token.allowance(yourAddress, custodyAddress)`
   - If insufficient: `token.approve(custodyAddress, amount)`

5. **Submit On-Chain Transaction**
   - Call `Custody.create(channel, state)` with both signatures
   - Wait for transaction confirmation

6. **Monitor for Channel Opening**

>    - Listen for `Opened` event (emitted when <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">the clearnode</Tooltip> joins)

Reviewer: In the current version of the protocol this event should be emitted right after the transaction in the previous step is mined

Author: [8211a936738924e9e67a3e56916ec7c6fee618be](https://github.com/layer-3/docs/pull/69/commits/8211a936738924e9e67a3e56916ec7c6fee618be)

   - Or poll `get_channels` until status is "active"

7. **Channel Active**
   - Channel appears in <Tooltip content="An abstraction that aggregates a user's funds across multiple blockchain networks, managed by Clearnode">unified balance</Tooltip>
   - Can now perform off-chain operations

> ### Error Cases

Reviewer: Currently we don't have error codes

Author: [8211a936738924e9e67a3e56916ec7c6fee618be](https://github.com/layer-3/docs/pull/69/commits/8211a936738924e9e67a3e56916ec7c6fee618be)

| **Clearnode unavailable** | -32603 | Clearnode at capacity | Retry later or contact support |

### Implementation Notes

- The <Tooltip content="Unique number ensuring channel identifier uniqueness">nonce</Tooltip> is generated by <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">the clearnode</Tooltip> to ensure uniqueness
- The channelId can be computed client-side: `keccak256(abi.encode(channel))`
- The <Tooltip content="Cryptographic hash of a state, used for signature verification">stateHash</Tooltip> should be verified: `keccak256(abi.encode(channelId, state.intent, state.version, state.data, state.allocations))`
- Users should verify <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">the clearnode's</Tooltip> signature before proceeding
- The challenge period can be customized but most users should use defaults

### Sequence Diagram

> ```mermaid

Reviewer: render error

Author: [8211a936738924e9e67a3e56916ec7c6fee618be](https://github.com/layer-3/docs/pull/69/commits/8211a936738924e9e67a3e56916ec7c6fee618be)

:::

### Prerequisites

- <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates">Channel</Tooltip> must exist and be in ACTIVE status
- User must be authenticated
- User must have native currency for gas fees
- Both parties must agree on final allocations (implicitly, by <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">the clearnode</Tooltip> signing)

### Request

| Parameter | Type | Required | Description | Default | Format | Example | Notes |
|-----------|------|----------|-------------|---------|--------|---------|-------|
| `channel_id` | string | Yes | Identifier of the channel to close | - | 0x-prefixed hex string (32 bytes) | `"0xabcdef1234567890..."` | Source: From get_channels or stored after creation |

> | `allocations` | Allocation[] | No | Proposed final allocation of funds<br/><br/>**Structure (per allocation)**:<br/>• `destination` (wallet address) - Where funds should go<br/>• `token` (wallet address) - Token contract address<br/>• `amount` (string) - Amount in smallest unit | The clearnode will use current unified balance if not provided | See structure | `[{"destination": "0x742d35Cc...", "token": "0x2791Bca1...", "amount": "105000000"}, {"destination": "0x123456Cc...", "token": "0x2791Bca1...", "amount": "0"}]` | Total must not exceed channel capacity<br/>Validation: The clearnode verifies against unified balance |

Reviewer: There is `funds_destination` field instead of this one

Author: [8211a936738924e9e67a3e56916ec7c6fee618be](https://github.com/layer-3/docs/pull/69/commits/8211a936738924e9e67a3e56916ec7c6fee618be)

### Response

| Parameter | Type | Description | Format/Structure | Example | Notes |
|-----------|------|-------------|------------------|---------|-------|

> | `channel` | Channel | Channel configuration (same as creation) | - | - | For verification and on-chain submission |

Reviewer: there is no such field in response params

Author: [8211a936738924e9e67a3e56916ec7c6fee618be](https://github.com/layer-3/docs/pull/69/commits/8211a936738924e9e67a3e56916ec7c6fee618be)

| `state` | State | Final state with CHANCLOSE magic number and the clearnode's signature<br/><br/>**Structure**:<br/>• `data` (bytes) - Contains CHANCLOSE magic number (7879 / 0x1EC7)<br/>&nbsp;&nbsp;- Purpose: Signals this is a final state for closure<br/>• `allocations` (Allocation[]) - Final allocation of funds<br/>&nbsp;&nbsp;- Note: Reflects current balances after all off-chain activity<br/>&nbsp;&nbsp;- Validation: Sum matches channel total funds<br/>• `sigs` (Signature[]) - Array of signatures; the clearnode's at index 1<br/>&nbsp;&nbsp;- Note: User must add their signature at index 0 | See structure | `data: "0x0000...1ec7"`<br/>`allocations: [{"destination": "0x742d35Cc...", "token": "0x2791Bca1...", "amount": "105000000"}, {"destination": "0x123456Cc...", "token": "0x2791Bca1...", "amount": "0"}]`<br/>`sigs: [null, "0xClearnodeFinalSignature..."]` | - |

> | `broker_signature` | string | The clearnode's signature over the final state | 0x-prefixed hex string (65 bytes) | `"0xabcdef987654..."` | Proves the clearnode agrees to this closure |

Reviewer: ```suggestion
| `server_signature` | string | The clearnode's signature over the final state | 0x-prefixed hex string (65 bytes) | `"0xabcdef987654..."` | Proves the clearnode agrees to this closure |
```

Author: [8211a936738924e9e67a3e56916ec7c6fee618be](https://github.com/layer-3/docs/pull/69/commits/8211a936738924e9e67a3e56916ec7c6fee618be)

   - Compute `stateHash` = `keccak256(abi.encode(channelId, state.intent, state.version, state.data, state.allocations))`
   - Verify signature is from <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">the clearnode</Tooltip>

3. **Sign Final State**
   - Sign `stateHash` with your <Tooltip content="An entity (identified by a wallet address) that is part of a channel">participant</Tooltip> key
   - Add signature at `state.sigs[0]`

4. **Submit On-Chain**
   - Call `Custody.close(channelId, state, proofs)` on blockchain
   - Both signatures must be present

5. **Wait for Confirmation**
   - Transaction confirms
   - Funds distributed according to allocations

6. **Channel Closed**
   - <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates">Channel</Tooltip> deleted from chain
   - Funds in your wallet or custody available balance

7. **Withdraw if Needed**
   - If funds in custody, call `withdraw()` to move to wallet

> ### Error Cases

Reviewer: Currently we don't have error codes

Author: [8211a936738924e9e67a3e56916ec7c6fee618be](https://github.com/layer-3/docs/pull/69/commits/8211a936738924e9e67a3e56916ec7c6fee618be)

:::tip In-Place Update
The resize operation updates the channel **in place**. The <Tooltip content="A unique identifier for a channel, formatted as a 0x-prefixed hex string (32 bytes)">channelId</Tooltip> **stays the same**, and the channel remains ACTIVE throughout. This is the current implementation of channel capacity adjustment.
:::

### Prerequisites

- <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates">Channel</Tooltip> must exist and be in ACTIVE status
- User must be authenticated
- For **positive delta** (deposit): User must have sufficient additional tokens
- For **negative delta** (withdrawal): Channel must have sufficient allocated funds
- All <Tooltip content="An entity (identified by a wallet address) that is part of a channel">participants</Tooltip> must sign the resize state (consensus required)
- User must have native currency for gas fees

> ### Request

Reviewer: Why not table as in the previous sections?

Author: [8211a936738924e9e67a3e56916ec7c6fee618be](https://github.com/layer-3/docs/pull/69/commits/8211a936738924e9e67a3e56916ec7c6fee618be)

    Required: Yes
    Format: 0x-prefixed hex string (32 bytes)
    Example: "0xabcdef1234567890..."
    Note: This channel_id will NOT change after resize
    
  allocate_amount: string
    Description: Amount to allocate within the unified balance before resize
    Required: No
    Format: Decimal string in human-readable format
    Default: "0"
    Example: "50.0"
    Note: Used for complex multi-chain rebalancing scenarios
    
  resize_amount: string
    Description: Delta amount to add (positive) or remove (negative)
    Required: Yes
    Format: Decimal string in human-readable format (can be negative)
    Examples:
      - "75.0" (deposit 75 more)
      - "-100.0" (withdraw 100)
    Validation:
      - Positive: User must have this amount available
      - Negative: Channel must have at least this amount allocated
    
  funds_destination: string (wallet address)
    Description: Destination for withdrawn funds (if resize_amount is negative)
    Required: Only if resize_amount < 0
    Format: 0x-prefixed hex string (20 bytes)
    Example: "0x742d35Cc6634C0532925a3b844Bc9e7595f0bEb"
    Note: Typically the user's wallet address
}
```

> ### Response

Reviewer: Why not table as in the previous sections?

Author: [8211a936738924e9e67a3e56916ec7c6fee618be](https://github.com/layer-3/docs/pull/69/commits/8211a936738924e9e67a3e56916ec7c6fee618be)

    Structure:
      intent: 2
        Description: StateIntent.RESIZE
        Note: Signals this is a resize operation
        
      version: number
        Description: precedingState.version + 1
        Example: If current version is 5, this will be 6
        Note: Version increments like any state update
        
      data: bytes
        Description: ABI-encoded int256[] delta amounts
        Format: abi.encode(int256[])
        Example: 
          - For +50 USDC by user: [50000000, 0]
          - For -30 USDC by user: [-30000000, 0]
        Note: Deltas are per participant (user at index 0, clearnode at index 1)
        
      allocations: Allocation[]
        Description: FINAL allocation state after resize
        Example (after +50 USDC deposit by user):
          Before: [{dest: user, amount: 100000000}, {dest: clearnode, amount: 0}]
          After:  [{dest: user, amount: 150000000}, {dest: clearnode, amount: 0}]
        Note: These are absolute amounts, not deltas
        
      sigs: Signature[]
        Description: Array of signatures (clearnode's at index 1)
        Example: [null, clearnodeSignature]
        Note: User must add their signature at index 0
    

>   proofs: State[]

Reviewer: there is no such field there

Author: [8211a936738924e9e67a3e56916ec7c6fee618be](https://github.com/layer-3/docs/pull/69/commits/8211a936738924e9e67a3e56916ec7c6fee618be)

    Description: Array of preceding states to support the resize
    Purpose: Allows adjudicator to validate the resize follows a valid state
    Note: May be empty or contain preceding states depending on adjudicator
    

>   clearnode_signature: string

Reviewer: `server_signature`

Author: [8211a936738924e9e67a3e56916ec7c6fee618be](https://github.com/layer-3/docs/pull/69/commits/8211a936738924e9e67a3e56916ec7c6fee618be)

    Description: Clearnode's signature on the resize state
    Format: 0x-prefixed hex string (65 bytes)
    Example: "0x9876fedcba..."
    Use: Already included in state.sigs[1]
}
```

### Next Steps After Receiving Response

The client must submit the resize state to the blockchain:

1. **Verify the resize state**
   - Check channel_id matches (should be unchanged)
   - Verify intent is RESIZE (2)
   - Confirm version is current + 1
   - Check allocations reflect the requested change

2. **Sign the resize state**
   ```
   stateHash = keccak256(abi.encode(
     channel_id,
     state.intent,      // StateIntent.RESIZE (2)
     state.version,     // Incremented version
     state.data,        // ABI-encoded int256[] deltas
     state.allocations  // Final allocations
   ))
   user_signature = sign(stateHash, participant_private_key)
   state.sigs[0] = user_signature
   // state.sigs[1] already has clearnode signature
   ```

> 3. **Approve additional token spending (if depositing)**

Reviewer: No, if depositing, then the requested amount should available in the custody ledger, so no need for approve

Author: [8211a936738924e9e67a3e56916ec7c6fee618be](https://github.com/layer-3/docs/pull/69/commits/8211a936738924e9e67a3e56916ec7c6fee618be)

   - Version incremented

### Error Cases

| Error | Cause | Resolution |
|-------|-------|------------|
| Authentication required | Not authenticated | Complete authentication flow |
| Channel not found | Invalid channel_id | Verify <Tooltip content="A unique identifier for a channel, formatted as a 0x-prefixed hex string (32 bytes)">channel ID</Tooltip> with `get_channels` |
| Channel not in ACTIVE status | Channel closed or in dispute | Cannot resize non-active <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates">channels</Tooltip> |
| Insufficient balance | Not enough available for deposit | Check `get_balance` for available amount |
| Exceeds channel capacity | Withdrawal exceeds allocated funds | Reduce withdrawal amount |
| Invalid signatures | Missing or incorrect signatures | All participants must sign resize state |
| Version conflict | State version doesn't match expected | Fetch current state, retry with correct version |
| Invalid proofs | Preceding states invalid | Ensure proofs chain correctly |

> ### Resize Scenarios

Reviewer: @nksazonov @dimast-x 
I think we should rewrite them to conform the current version of the protocol, where we expect empty channel allocations
